Add a bevy-texture example based on wgpu-texture

jerome-caucat opened the pull request:
In order to integrate Bevy with Dioxus, I adapted the `wgpu-texture` example to use Bevy for the 3D rendering.

I added a new `bevy-texture` example where:
- The rendering is done by a headless Bevy app.
- Each frame is captured as a screenshot and used to fill a wgpu texture.
  - This part could probably be improved.
- The texture is displayed using `dioxus_native` custom paint source.

[Screencast from 2025-07-03 15-22-42.webm](https://github.com/user-attachments/assets/ba0581cc-266d-4002-a240-322d8408a511)

closes https://github.com/DioxusLabs/dioxus/issues/4359

nicoburns commented:
I think we might want to use a https://docs.rs/bevy/latest/bevy/render/camera/struct.ManualTextureView.html as the render target.

jerome-caucat commented:
> I think we might want to use a https://docs.rs/bevy/latest/bevy/render/camera/struct.ManualTextureView.html as the render target (or an `Image`, but access the texture directly if that's possible).

@nicoburns Yes, I am going to fix that. Thanks for the suggestion!

I also would like to add a simple camera controller, to showcase how to pass mouse and key events to Bevy. Not sure how I am going to transfer the events yet.

nicoburns commented:
> Perhaps the problem is that I am using the device from Dioxus to create the texture.

Yes, I had the same problem with the `wgpu-texture` example when I was using two separate devices. I think that one way or the other you'll need Bevy and Blitz to use the same `wgpu::Device`

nicoburns commented:
> I had a quick look at [RenderCreation::Manual](https://docs.rs/bevy/latest/bevy/render/settings/enum.RenderCreation.html) but it seems hard to setup so that the Bevy app works normally. I'll give it another go.

Does this help? https://github.com/DioxusLabs/blitz/pull/232 (you would need to point `dioxus-native` to the git version of Blitz)

I think that gives you access to all the types you need. `Instance` is passed directly, `DeviceHandle` contains `Device`, `Queue`, and `Adapter` (you can clone them if you need owned versions), and `AdapterInfo` can be obtained from `Adapter`.

nicoburns commented:
@jerome-caucat Ok, I have [RenderCreation::Manual](https://docs.rs/bevy/latest/bevy/render/settings/enum.RenderCreation.html) setup. So you should be able to try the texture code again now.

nicoburns commented:
Ok, and I also ported the code you [posted above](https://github.com/DioxusLabs/dioxus/pull/4360#issuecomment-3036206658) (sorry - I got excited!).

Seems to be working! I'm not sure if it is correctly responding to resizes: the `wgpu-texture` cube seems to resize to match the window size whereas bevy code does not (not sure if this is bug where the resize isn't being picked up, or just a difference in how the Bevy scene is being rendered)

jerome-caucat commented:
@nicoburns Thanks a lot for your help. I was also updating the code on my side. I'll clean the code and fix the remaining issues.

nicoburns commented:
I've published a new versions of the `anyrender` and `blitz` crates so this PR doesn't need to depend on git versions. Anyrender crates should be set to `0.4.0`. Blitz crates to `0.1.0-alpha.5`

jerome-caucat commented:
@nicoburns I cleaned the code, added some comments, fixed the resize problem and renamed the example.

I noted there is a segfault when we close the window, but the `wgpu-texture` example has the same issue.

I wanted to add mouse and keyboard events to this demo, but perhaps I should do this in a separate issue.

jerome-caucat commented on examples/bevy-texture/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "bevy-texture"
```
@nicoburns Done.

nicoburns reviewed (approved):
@jerome-caucat All looks good to me. Just need to fix the issues with CI and then we can land this. Long-term we may want to move this into a separate repo (as this one is already a very large workspace, and pulling in the whole of Bevy just for an example is quite a lot). But lets land it here first, and we move it and improve on it later.

> I noted there is a segfault when we close the window, but the wgpu-texture example has the same issue.

Not ideal, but not disastrous in an example. If the `wgpu-texture` example has the same problem let's not block this PR on it.

> I wanted to add mouse and keyboard events to this demo, but perhaps I should do this in a separate issue.

Yeah, let's do a new issue/PR. This will likely require further changes in the Blitz layer.
